Fix test stability (index not ready)

We are often seeing that kind of errors when running our tests from TravisCI:

```
  2> mvn integration-test -Dtests.seed=B2B81FA82F956189 -Dtests.class=fr.pilato.elasticsearch.crawler.fs.test.integration.FsCrawlerRestIT -Dtests.method="testDocumentWithExternalTags" -Dtests.locale=es-AR -Dtests.timezone=Atlantic/Reykjavik
ERROR   5.97s | FsCrawlerRestIT.testDocumentWithExternalTags <<<
   > Throwable #1: ElasticsearchStatusException[Elasticsearch exception [type=search_phase_execution_exception, reason=all shards failed]]
   > 	at __randomizedtesting.SeedInfo.seed([B2B81FA82F956189:1D2BC9C1D3840356]:0)
   > 	at org.elasticsearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:177)
   > 	at org.elasticsearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:2050)
   > 	at org.elasticsearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:2026)
   > 	at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1775)
   > 	at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1732)
   > 	at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1694)
   > 	at org.elasticsearch.client.RestHighLevelClient.search(RestHighLevelClient.java:1090)
   > 	at fr.pilato.elasticsearch.crawler.fs.client.v6.ElasticsearchClientV6.search(ElasticsearchClientV6.java:474)
   > 	at fr.pilato.elasticsearch.crawler.fs.test.integration.AbstractITCase.lambda$countTestHelper$1(AbstractITCase.java:376)
   > 	at fr.pilato.elasticsearch.crawler.fs.test.framework.AbstractFSCrawlerTestCase.awaitBusy(AbstractFSCrawlerTestCase.java:149)
   > 	at fr.pilato.elasticsearch.crawler.fs.test.integration.AbstractITCase.countTestHelper(AbstractITCase.java:371)
   > 	at fr.pilato.elasticsearch.crawler.fs.test.integration.FsCrawlerRestIT.testDocumentWithExternalTags(FsCrawlerRestIT.java:124)
   > 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   > 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   > 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   > 	at java.lang.reflect.Method.invoke(Method.java:498)
   > 	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1737)
   > 	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:934)
   > 	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:970)
   > 	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:984)
   > 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
   > 	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   > 	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
   > 	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
   > 	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
   > 	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:943)
   > 	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:829)
   > 	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:879)
   > 	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:890)
   > 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
   > 	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   > 	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
   > 	at java.lang.Thread.run(Thread.java:748)
   > 	Suppressed: org.elasticsearch.client.ResponseException: method [POST], host [http://127.0.0.1:9200], URI [/fscrawler_fs_crawler_rest_i_t_test_document_with_external_tags/_search?typed_keys=true&ignore_unavailable=true&expand_wildcards=open&allow_no_indices=true&search_type=query_then_fetch&batched_reduce_size=512], status line [HTTP/1.1 503 Service Unavailable]
   > {"error":{"root_cause":[],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[]},"status":503}
   > 		at org.elasticsearch.client.RestClient$SyncResponseListener.get(RestClient.java:929)
   > 		at org.elasticsearch.client.RestClient.performRequest(RestClient.java:229)
   > 		at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1762)
   > 		... 29 more
   > 	Caused by: org.elasticsearch.client.ResponseException: method [POST], host [http://127.0.0.1:9200], URI [/fscrawler_fs_crawler_rest_i_t_test_document_with_external_tags/_search?typed_keys=true&ignore_unavailable=true&expand_wildcards=open&allow_no_indices=true&search_type=query_then_fetch&batched_reduce_size=512], status line [HTTP/1.1 503 Service Unavailable]
   > {"error":{"root_cause":[],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[]},"status":503}
   > 		at org.elasticsearch.client.RestClient$1.completed(RestClient.java:548)
   > 		at org.elasticsearch.client.RestClient$1.completed(RestClient.java:533)
```

Probably we need to be more conservative when the index is created and wait for green or yellow status before actually indexing documents.

--- a/elasticsearch-client/elasticsearch-client-v5/src/main/java/fr/pilato/elasticsearch/crawler/fs/client/v5/ElasticsearchClientV5.java
+++ b/elasticsearch-client/elasticsearch-client-v5/src/main/java/fr/pilato/elasticsearch/crawler/fs/client/v5/ElasticsearchClientV5.java
@@ -50,7 +50,6 @@ import org.apache.logging.log4j.Logger;
 import org.elasticsearch.ElasticsearchStatusException;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionListener;
-import org.elasticsearch.action.admin.cluster.health.ClusterHealthRequest;
 import org.elasticsearch.action.bulk.BulkProcessor;
 import org.elasticsearch.action.bulk.BulkRequest;
 import org.elasticsearch.action.bulk.BulkResponse;

--- a/elasticsearch-client/elasticsearch-client-v5/src/main/java/fr/pilato/elasticsearch/crawler/fs/client/v5/ElasticsearchClientV5.java
+++ b/elasticsearch-client/elasticsearch-client-v5/src/main/java/fr/pilato/elasticsearch/crawler/fs/client/v5/ElasticsearchClientV5.java
@@ -50,6 +50,7 @@ import org.apache.logging.log4j.Logger;
 import org.elasticsearch.ElasticsearchStatusException;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.cluster.health.ClusterHealthRequest;
 import org.elasticsearch.action.bulk.BulkProcessor;
 import org.elasticsearch.action.bulk.BulkRequest;
 import org.elasticsearch.action.bulk.BulkResponse;
@@ -220,8 +221,7 @@ public class ElasticsearchClientV5 implements ElasticsearchClient {
             logger.trace("create index response: {}", asMap(response));
         } catch (ResponseException e) {
             if (e.getResponse().getStatusLine().getStatusCode() == 400 &&
-                    (e.getMessage().contains("index_already_exists_exception") || // ES 5.x
-                            e.getMessage().contains("IndexAlreadyExistsException") )) { // ES 1.x and 2.x
+                    (e.getMessage().contains("index_already_exists_exception"))) {
                 if (!ignoreErrors) {
                     throw new RuntimeException("index already exists");
                 }
@@ -230,6 +230,7 @@ public class ElasticsearchClientV5 implements ElasticsearchClient {
             }
             throw e;
         }
+        waitForHealthyIndex(index);
     }
 
     /**

--- a/elasticsearch-client/elasticsearch-client-v6/src/main/java/fr/pilato/elasticsearch/crawler/fs/client/v6/ElasticsearchClientV6.java
+++ b/elasticsearch-client/elasticsearch-client-v6/src/main/java/fr/pilato/elasticsearch/crawler/fs/client/v6/ElasticsearchClientV6.java
@@ -240,7 +240,11 @@ public class ElasticsearchClientV6 implements ElasticsearchClient {
             if (e.getMessage().contains("resource_already_exists_exception") && !ignoreErrors) {
                 throw new RuntimeException("index already exists");
             }
+            if (!e.getMessage().contains("resource_already_exists_exception")) {
+                throw e;
+            }
         }
+        waitForHealthyIndex(index);
     }
 
     /**

--- a/elasticsearch-client/elasticsearch-client-v7/src/main/java/fr/pilato/elasticsearch/crawler/fs/client/v7/ElasticsearchClientV7.java
+++ b/elasticsearch-client/elasticsearch-client-v7/src/main/java/fr/pilato/elasticsearch/crawler/fs/client/v7/ElasticsearchClientV7.java
@@ -239,7 +239,7 @@ public class ElasticsearchClientV7 implements ElasticsearchClient {
                 throw e;
             }
         }
-        client.cluster().health(new ClusterHealthRequest(index).waitForYellowStatus(), RequestOptions.DEFAULT);
+        waitForHealthyIndex(index);
     }
 
     /**

